daemon/compress.c

```diff
@@ -0,0 +1,75 @@
+
+#include <str.h>
+#include <zlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include "log.h"
+#include "udp_listener.h"
+#include "str.h"
+#include "compress.h"
+
+#define windowBits 15
+
+#define GZIP_ENCODING 16
+#define ENABLE_ZLIB_GZIP 32
+
+static void init_stream(z_stream *stream);
+
+static void init_stream(z_stream *stream)
+{
+	stream->zalloc = Z_NULL;
+	stream->zfree = Z_NULL;
+	stream->opaque = Z_NULL;
+}
+
+
+// compress_data compress the data of the given length.
+// returns compressed buffer length and replace the original buffer.
+int compress_data(char *buf, int len) {
```

Comment:
I don't think it's of very good form to use the same buffer for input and output, because we end up needing a temporary that we then need to copy, which looks inefficient.

Reply:
If you are saying so, I'd fix it.

But to return the compressed data it required allocating the memory using malloc()/calloc, and need to be freed them afterward. I thought that one is more expensive.

Before fixing it, I would like to listen to your opinion again.

Reply:
I don't think there's a need for memory allocation here, I think it would just be a matter of moving this around. You can just use another buffer in the call site, and pass it to the function. If we are doing compression then you can simply assign the output buffer to be passed to the send function, otherwise we pass the input one. This avoid allocation and an additional buffer copy, as we only need to copy a pointer instead.

Reply:
Np, I've got that. Fixed it. 👍